pallets/watchtower/src/lib.rs

```diff
@@ -0,0 +1,1195 @@
+                        *vote_is_valid,
+                        current_block,
+                        source,
+                        signature.encode()[0..8].to_vec(),
+                    );
+
+                    let provides_tag = unique_payload_for_provides.encode();
+
+                    log::info!(
+                        target: "runtime::watchtower::validate",
+                        "[VALIDATE_UNSIGNED] ACCEPTED for node {:?} at block {:?}. Source: {:?}. Provides tag hash: {:?}. Vote: {:?}",
+                        node, current_block, source, sp_io::hashing::blake2_256(&provides_tag), vote_is_valid
+                    );
+
+                    ValidTransaction::with_tag_prefix("WatchtowerOCW")
+                        .priority(TransactionPriority::MAX)
+                        .and_provides(vec![provides_tag])
+                        .longevity(64_u64)
+                        .propagate(true)
+                        .build()
+                } else {
+                    log::error!(
+                        target: "runtime::watchtower::validate",
+                        "[VALIDATE_UNSIGNED] REJECTED (Signature Invalid): Invalid signature for node {:?}.",
+                        node
+                    );
+                    InvalidTransaction::BadSigner.into()
+                }
+            } else {
+                log::warn!(
+                    target: "runtime::watchtower::validate",
+                    "[VALIDATE_UNSIGNED] WARNING: Received non-watchtower-vote call for unsigned validation: {:?}.",
+                    call
+                );
+                InvalidTransaction::Call.into()
+            }
+        }
+    }
+
+    impl<T: Config> Pallet<T> {
+        pub fn notify_summary_ready_for_validation(
+            instance: SummarySourceInstance,
+            root_id: WatchtowerRootId<BlockNumberFor<T>>,
+            root_hash: WatchtowerOnChainHash,
+        ) -> DispatchResult {
+            log::info!(
+                target: "runtime::watchtower::notification",
+                "Received notification: instance={:?}, root_id={:?}, hash={:?}",
+                instance, root_id, root_hash
+            );
+
+            let consensus_key = (instance, root_id.clone());
+
+            if VoteConsensusReached::<T>::get(&consensus_key) {
+                log::warn!(
+                    target: "runtime::watchtower::notification",
+                    "Consensus already reached for summary: {:?}",
+                    consensus_key
+                );
+                return Ok(());
+            }
+
+            if VotingStartBlock::<T>::get(&consensus_key).is_some() {
+                log::warn!(
+                    target: "runtime::watchtower::notification",
+                    "Voting already active for summary: {:?}",
+                    consensus_key
+                );
+                return Ok(());
+            }
+
+            if root_hash == sp_core::H256::zero() {
```

Comment:
should we allow empty root hashes? If No, we should error here. If YES we should not have the `if` statement.